clients/keycloak/src/main/kotlin/DefaultKeycloakClient.kt

```diff
@@ -153,12 +155,36 @@ class DefaultKeycloakClient(
         internalClientId ?: findClientId(clientId).also { internalClientId = it }
     }
 
-    override suspend fun getGroups(): Set<Group> =
+    override suspend fun getGroups(groupNameFilter: String?): Set<Group> =
         runCatching {
-            httpClient.get("$apiUrl/groups")
+            val groupCount = httpClient.get("$apiUrl/groups/count") {
```

Comment:
Idea for a further optimization: Is it possible to drop this call to get the group count by counting the number of groups returned for a chunk? If a request for a chunk of groups yields less than `GROUP_GET_CHUNK_SIZE` elements, the last chunk has been reached.

Reply:
Note that this is the *total* number of groups, which is used to optimize the capacity for the set in line 168. Esp. for a large number of groups / chunks properly setting the initial capacity can also yield significant performance gains. I'm not sure whether the same would work with your approach.